src/Channel.cpp

```diff
@@ -84,7 +87,85 @@ void Channel::remove(Channel *c) {
 	QWriteLocker lock(&c_qrwlChannels);
 	c_qhChannels.remove(c->iId);
 }
-#endif
+
+void Channel::setFilterMode(Channel::FilterMode filterMode) {
+	m_filterMode = filterMode;
+	ServerHandlerPtr sh = Global::get().sh;
+	if (sh) {
+		Global::get().db->setChannelFilterMode(sh->qbaDigest, iId, m_filterMode);
+	}
+}
+
+bool Channel::isFiltered() const {
+	if (!Global::get().s.bFilterActive) {
+		// Channel filtering is disabled
+		return false;
+	}
+
+	Channel *userChannel = nullptr;
+
+	if (Global::get().uiSession != 0) {
+		const ClientUser *user = ClientUser::get(Global::get().uiSession);
+		if (user) {
+			userChannel = user->cChannel;
+		}
+	}
+
+	bool hasUser = false;
+	bool hasHide = false;
+	bool hasPin  = false;
+
+	// Iterate tree down
+	std::queue< const Channel * > channelSearch;
+	channelSearch.push(this);
+
+	while (!channelSearch.empty()) {
+		const Channel *c = channelSearch.front();
+		channelSearch.pop();
+
+		// Never hide channel, if user resides in this channel or a subchannel
+		if (userChannel != nullptr && c == userChannel) {
+			return false;
+		}
+
+		if (!c->qlUsers.isEmpty()) {
+			hasUser = true;
+		}
+
+		if (c->m_filterMode == Channel::FILTER_MODE_PIN) {
+			hasPin = true;
+		}
+
+		for (int i = 0; i < c->qlChannels.count(); i++) {
+			channelSearch.push(c->qlChannels[i]);
+		}
+	}
+
+	// Iterate tree up
+	const Channel *c = this;
+	while (c) {
+		if (c->m_filterMode == Channel::FILTER_MODE_HIDE) {
+			hasHide = true;
```

Comment:
As above: We can end the loop as soon as we encountered this
```suggestion
		if (c->m_filterMode == Channel::FILTER_MODE_HIDE) {
			hasHide = true;
			break;
```